### PR TITLE
fix POST/PUT/PATCH has no body redbox, when xhr is used without body

### LINKED ISSUE #3371: [Network] Android throws up a redbox when POST has no body
This occurs when you make a call like `fetch(url, { method: 'post' })`. The native layer should fill in an empty body instead of fataling.

### PR DESCRIPTION
**Pull request by qbig**
… by passing a empty body
fix #3371
referring to https://github.com/square/okhttp/pull/1559/files

### CODE REVIEW

File: ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java [hunk 1]
```diff
@@ -136,7 +135,13 @@ public void sendRequest(
     requestBuilder.headers(requestHeaders);
 
     if (data == null) {
-      requestBuilder.method(method, null);
+      if(method.equals("POST")
```
Comment (satya164) [inline ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java]: Space after `if`. Also I think the condition can go in one line.

**Comment by ajwhite** [comment]
Looks like the build could use a restart, few npm installation failures

File: ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java [hunk 2]
```diff
@@ -136,7 +135,11 @@ public void sendRequest(
     requestBuilder.headers(requestHeaders);
 
     if (data == null) {
-      requestBuilder.method(method, null);
+      if (method.equals("POST") || method.equals("PUT") || method.equals("PATCH")){
+        requestBuilder.method(method, RequestBody.create(null, new byte[0]));
```
Comment (lexs) [inline ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java]: You could use ByteString.EMPTY here.

File: ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java [hunk 3]
```diff
@@ -136,7 +135,11 @@ public void sendRequest(
     requestBuilder.headers(requestHeaders);
 
     if (data == null) {
-      requestBuilder.method(method, null);
+      if (method.equals("POST") || method.equals("PUT") || method.equals("PATCH")){
```
Comment (lexs) [inline ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java]: Can you put this in a method, like `getEmptyBody(String method)` and put it in `RequestBodyUtil`. Then you'd just have:
    requestBuilder.method(method, getEmptyBody(method));

Comment (qbig) [inline ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java]: @lexs That's really neat! Would update soon.

File: ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java (hunk 2 above)
Comment (qbig) [inline ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java]: Are they entirely same as in 
```java
public static final ByteString EMPTY = new ByteString(new byte[0]);
```
from https://github.com/google/nfc-smart-tag/blob/master/demoserver/src/com/google/protobuf/ByteString.java

**Comment by mkonicek** [comment]
Thanks for the review @lexs!

File: ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java (hunk 2 above)
Comment (qbig) [inline ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java]: @lexs  could you explain or give me a pointer? Thanks:)

Comment (qbig) [inline ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java]: @mkonicek Could you clarify why @lexs or how I could use ByteString.EMPTY here instead ?

Comment (lexs) [inline ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java]: @qbig I was referring to the Okio `ByteString`: https://github.com/square/okio/blob/master/okio/src/main/java/okio/ByteString.java#L52

Which you can pass to `RequestBody.create()`: https://github.com/square/okhttp/blob/master/okhttp/src/main/java/com/squareup/okhttp/RequestBody.java#L60

Comment (qbig) [inline ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java]: arrr. I see. Thanks.

Comment (qbig) [inline ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java]: @lexs updated:D

**Comment by qbig** [comment]
updated:)

File: ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java [hunk 4]
```diff
@@ -9,12 +9,6 @@
 
 package com.facebook.react.modules.network;
 
-import javax.annotation.Nullable;
-
-import java.io.IOException;
-import java.io.InputStream;
-import java.io.Reader;
-
```
Comment (lexs) [inline ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java]: Could you make sure the imports doesn't move here. There should be an option in your IDE. The problem is that another PR to this file will probably end moving them back destroying history again.

Same of the other file.

Comment (qbig) [inline ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java]: :) updated!

**Comment by lexs** [comment]
@facebook-github-bot shipit

**Comment by sreesharp** [comment]
Hi @qbig, I am still getting this error with  react-native - v0.19.0 on Android.  Let me know if this fix is available in production version.

**Pull request closed without merging**